### DISCUSSION TITLE
How does rbenv know where to find packages?

### DISCUSSION THREAD

**Discussion by AdrianTP** [rbenv/rbenv#1529] [Q&A] [answered]
When I run `rbenv install <version>`, how does rbenv generate the https://cache.ruby-lang.org/pub/ruby/ url from which it downloads the associated ruby package? A client is looking to start caching binaries and source packages in a tool like JFrog, which for something like Gems means simply modifying the Gemfile to point at the correct repo URL, but I don't see an option like that for rbenv.

I have tried to dissuade them from this course of action but nobody will hear me, so I need to start figuring out how to play along and get ahead of the tidal wave by modifying our environment setup scripts and documentation to adapt to this paradigm shift. Thanks!

**Comment by mislav** [comment, marked as answer]
I'm not completely sure about your question re: gems and Gemfiles, but I can answer about how rbenv knows where to download the Ruby package from:

1. `rbenv install 3.1.4` calls `ruby-build 3.1.4 ~/.rbenv/versions/3.1.4`;
2. Within the ruby-build project, that activates the Ruby [`3.1.4` definition file](https://github.com/rbenv/ruby-build/blob/master/share/ruby-build/3.1.4);
3. That definition file specifies that Ruby should be downloaded from `https://cache.ruby-lang.org/pub/ruby/3.1/ruby-3.1.4.tar.gz` and compiled;
4. Optionally, the RUBY_BUILD_MIRROR_URL environment variable can be set to download ruby-build packages from an alternate host, e.g. an internal mirror server. [See here for more information.](https://github.com/rbenv/ruby-build#package-mirrors)

However, when Gems are installed within the context of a Ruby version, that has nothing to do with rbenv/ruby-build. Instead, the Gems download location is determined by Rubygems or Bundler. They might have some features for fetching packages or binaries from an alternate host, but I'm not familiar with the specifics.

**Reply by AdrianTP** [reply]
This is exactly what I was looking for. Thank you!
